File: src/dataset/python/meetup_exercism-ThomasZumsteg/claude/ClaudeSonnet4_meetup_v3.py
"""Returns a datetime date object for a given year, month, and weekday that meet
a predefined natrual language description. e.g. 1st, 2nd, last, teenth
"""

import datetime
import calendar

WEEKDAY_TO_NUM = {
    'Monday': 0, 'Tuesday': 1, 'Wednesday': 2, 'Thursday': 3,
    'Friday': 4, 'Saturday': 5, 'Sunday': 6
}

def meetup_day(year, month, weekday, condition):
    """Returns date object for a given year, month, weekday and given condition
    Current conditions [1-4]th occurence, last, teenth
    """
    target_weekday = WEEKDAY_TO_NUM[weekday]
    
    if condition == 'teenth':
        for day in range(13, 20):
            date_obj = datetime.date(year, month, day)
            if date_obj.weekday() == target_weekday:
                return date_obj
    
    elif condition == 'last':
        last_day = calendar.monthrange(year, month)[1]
        for day in range(last_day, 0, -1):
            date_obj = datetime.date(year, month, day)
            if date_obj.weekday() == target_weekday:
                if (date_obj + datetime.timedelta(days=7)).month != month:
                    return date_obj
    
    elif condition in ('1st', '2nd', '3rd', '4th', '5th'):
        occurrence = int(condition[0])
        count = 0
        for day in range(1, calendar.monthrange(year, month)[1] + 1):
            date_obj = datetime.date(year, month, day)
            if date_obj.weekday() == target_weekday:
                count += 1
                if count == occurrence:
                    return date_obj
    
    raise AssertionError("Not a valid meetup day")

def last(day):
    """True on the last week of the month"""
    return day.month != (day + datetime.timedelta(days=7)).month

def nth(n):
    """Returns a function that is true on the nth week of a month"""
    def nth_worker(day):
        """True on the nth week of the month"""
        return (day.day - 1) // 7 == n-1
    return nth_worker

def teenth(day):
    """True on a days ending in -teenth, ie seventeenth"""
    return 13 <= day.day <= 19

CONDITIONS = {
    'teenth': teenth,
    'last': last,
    '1st': nth(1),
    '2nd': nth(2),
    '3rd': nth(3),
    '4th': nth(4),
    '5th': nth(5),
}

File: src/dataset/javascript/meetup_exercism-ThomasZumsteg/claude/ClaudeSonnet4_meetup_v3.js
const weekDays = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];

const nthRanges = {
  "teenth": [13, 19],
  "1st": [1, 7],
  "2nd": [8, 14],
  "3rd": [15, 21],
  "4th": [22, 28],
  "5th": [29, 31]
};

var Meetup = function(year, month, day, nth) {
  const intDay = weekDays.indexOf(day);
  
  if (nth === "last") {
    return last(year, month, intDay);
  }
  
  const range = nthRanges[nth];
  if (range) {
    return find_day(year, month, intDay, range[0], range[1]);
  }
  
  throw "Invalid nth parameter";
};

function find_day(year, month, day, start, stop) {
  const startDate = new Date(year, month, start);
  const targetDay = day;
  const currentDay = startDate.getDay();
  
  let daysToAdd = (targetDay - currentDay + 7) % 7;
  let meetupDate = start + daysToAdd;
  
  if (meetupDate <= stop) {
    const result = new Date(year, month, meetupDate);
    if (result.getMonth() === month) {
      return result;
    }
  }
  
  meetupDate += 7;
  if (meetupDate <= stop) {
    const result = new Date(year, month, meetupDate);
    if (result.getMonth() === month) {
      return result;
    }
  }
  
  throw "Date does not exist";
}

function last(year, month, day) {
  const lastDayOfMonth = new Date(year, month + 1, 0);
  const lastDate = lastDayOfMonth.getDate();
  const lastDay = lastDayOfMonth.getDay();
  
  let daysToSubtract = (lastDay - day + 7) % 7;
  const meetupDate = lastDate - daysToSubtract;
  
  return new Date(year, month, meetupDate);
}

Date.prototype.addDays = function(days) {
  const dat = new Date(this.valueOf());
  dat.setDate(dat.getDate() + days);
  return dat;
};

module.exports = Meetup;